### PR TITLE
Adding deepMap implementation

### PR DESCRIPTION
**Pull request by dkzlv**
TLDR: MapStore that's friendly to nested attributes and arrays.

Main magic happens in our own naïve implementation of object-path getting/setting. It uses recursion to only change object reference (via `structuredClone`) for the object that has some changed keys. It only works with plain objects and arrays, not sets, maps, getters/setters and other stuff.

As for tests, I copied most of them from `map`, but adjusted them/added new for nested keys scenarios.

Yet to do:

- [ ] documentation
- [ ] update react bindings so that `opt.keys` is statically typed for `deepMap`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 node_modules/
 
 coverage/
+.vscode
```
Comment (ai) [inline .gitignore:4]: IDE configs should be in your local git ignore file

File: deep-map/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+export type DeepMapStore<T extends Record<string, unknown>> = Omit<
+  WritableAtom<T>,
+  'setKey' | 'listen' | 'subscribe'
+> & {
+  setKey: <K extends AllKeys<T>>(key: K, value: GetPath<T, K>) => void
+  listen: Listener<T>
+  subscribe: Listener<T>
+}
+
+export function deepMap<T extends Record<string, unknown>>(
```
Comment (ai) [inline deep-map/index.d.ts:19]: We need TypeDocs here

File: deep-map/path.d.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import type { Call, Objects } from 'hotscript'
```
Comment (ai) [inline deep-map/path.d.ts:1]: We will need to add `hotscript` to project’s `dependencies`

File: deep-map/path.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
+/**
+ * Get a deep value by key. Undefined if key is missing.
+ * Doesn't have a lot of runtime sanity checks, instead relies on TS to tell the user about
+ * their mistakes.
+ *
+ * @param obj Any object you want to get a deep path of
+ * @param path Path splitted by dots. Arrays accessed the same as in JS: props.arr[1].nested
+ */
+export const getPath = (obj, path) => {
```
Comment (ai) [inline deep-map/path.js:9]: We are using `export function X()` for these cases. Let’s keep consistency.

File: deep-map/index.d.ts (hunk 2 above)
Comment (ai) [reply deep-map/index.d.ts:19]: I like the idea. Some users already asked for something like this.

Comment (dkzlv) [reply deep-map/index.d.ts:63]: @ai Have a look. Added docs to all exported functions and to README itself.

File: deep-map/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
+import { WritableAtom } from '../atom/index.js'
+import { AllKeys, GetPath } from './path.js'
+
+export * from './path.js'
+
+type Listener<T extends Record<string, unknown>> = (
+  listener: (value: T, changedKey: undefined | AllKeys<T>) => void
+) => () => void
+
+export type DeepMapStore<T extends Record<string, unknown>> = Omit<
```
Comment (ai) [inline deep-map/index.d.ts:10]: Can we pass `deepMap` to argument for `atom` after omitting and replacing this methods?

File: deep-map/index.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import { atom } from '../atom/index.js'
+import { getPath, setPath } from './path.js'
+
+export * from './path.js'
```
Comment (ai) [inline deep-map/index.js:4]: Can we move this function here? Otherwise this file become too small 😅

File: pnpm-lock.yaml [hunk 7]
```diff
@@ -20,12 +20,16 @@ specifiers:
...
   eslint-plugin-promise: ^6.1.1
+  hotscript: ^1.0.8
   nanodelay: ^2.0.2
   size-limit: ^8.2.4
   tsm: ^2.3.0
   typescript: ^4.9.5
   uvu: ^0.5.6
 
+dependencies:
+  hotscript: 1.0.8
```
Comment (ai) [inline pnpm-lock.yaml:31]: Can we avoid dependency (for insurance by copying types helpers)?

Right now it is zero-dependency project which is one of the argument why it is minimalistic. I am ok to loose this feature for a big feature, but not here.

File: deep-map/index.js (hunk 6 above)
Comment (dkzlv) [reply deep-map/index.js:4]: Currently modules are separated by their function. I surely can merge them, but it'll both make type notations harder to read/edit and the module itself will be messier.

> file become too small

… so?

File: deep-map/index.d.ts (hunk 5 above)
Comment (dkzlv) [reply deep-map/index.d.ts:10]: Can you explain it once again? I didn't follow.

File: pnpm-lock.yaml (hunk 7 above)
Comment (ai) [reply pnpm-lock.yaml:31]: > What kind of insurance?

Can we avoid dependency (for *instance by copying types helpers)?

Comment (dkzlv) [reply pnpm-lock.yaml:31]: I mean, insurance from what exactly?

Comment (ai) [reply pnpm-lock.yaml:31]: No insurance, it was just typo.

Having extra dependency is loosing one of our feature. How many code do we use from dependency? Can we just copy it?

Comment (dkzlv) [reply pnpm-lock.yaml:31]: We use quite a lot of their code. These are just two type functions (`Objects.AllPaths` and `GetFromPath`), but those are tightly entangled with the rest of the lib, and for a good reason (making the same features without this entanglement is quite hard, I was talking about this in Slack).
If I were to copy everything those two function rely on I would copy at least 150-200 LOCs of code that I couldn't support in any way. And I can't really say if I can write those helpers myself in a meaningful time.

Still, I don't understand what you mean by "zero dependency". This is pure TS types, it won't bloat the bundle in any way. I would think "zero dependency" is mostly about code-emitting deps. We can rename the feature from "Zero dependencies" to "Zero runtime dependencies" 😂

Comment (dkzlv) [reply pnpm-lock.yaml:31]: On top of that, if we one day decide to move away from `.js` + `.d.ts` to plain `.ts`, `hotscript` would actually simplify most of the typings we have.

Comment (ai) [reply pnpm-lock.yaml:31]: > This is pure TS types, it won't bloat the bundle in any way

People will not try to understand what type of dependency do they see. They still will see a dependency and will think about the project that it is not so `nano-` anymore.

File: deep-map/index.d.ts (hunk 5 above)
Comment (ai) [reply deep-map/index.d.ts:10]: Does this code work:

```ts
import { deepMap, ReadableAtom } from 'nanostore'

function someStoreFunc (store: ReadableAtom) {
  …
}

someStoreFunc(deepMap())
```

Comment (dkzlv) [reply deep-map/index.d.ts]: Yeah, sure, it will. `ReadableAtom` has a very simple signature.

**Pull request merged**